### ISSUE TITLE
Not concurrent?

### ISSUE THREAD

**Issue by strohhut** [David-Desmaisons/RateLimiter#4]
It appears to me that RateLimiter runs the tasks in sequence. This can make the application very slow.

```[CSharp]
[Test]
public void RateLimiter()
{
    TimeLimiter timeconstraint = TimeLimiter.GetFromMaxCountByInterval(3, TimeSpan.FromSeconds(1));

    var tasks = new List<Task>();

    for (int i = 0; i < 15; i++)
    {
        tasks.Add(timeconstraint.Perform(ConsoleItAsync));
    }

    Task.WaitAll(tasks.ToArray());
}

private static async Task ConsoleItAsync()
{
    await Task.Delay(800);

    var id = Thread.CurrentThread.ManagedThreadId;

    Trace.WriteLine($"{id}: {DateTime.Now:MM/dd/yyy HH:mm:ss.fff}");
}
```

Here I'm limiting to 3 executions per second. But because `ConsoleItAsync` is so slow I get fewer than 3 executions per second:
```
4: 10.30.2017 13:26:09.392
4: 10.30.2017 13:26:10.203
4: 10.30.2017 13:26:11.012
```

I think it would be better to start three tasks in parallel and when the first tasks finishes check if there is room to start another task (e.g. maxCount and timespan allow it). Otherwise `Task.Delay` it.

Or maybe make it more explicit in the docs that the execution will be sequentially. Because I think that people expect concurrency when they start multiple tasks via `Perform`.

**Comment by David-Desmaisons** [comment]
Hello @strohhut, this is done this way exactly by design:

-RateLimiter when called with a Fun<Task> wait for completion of the task before running another one.

-If you need concurency, then use the version of perform that takes an action:

```[CSharp]
[Test]
public void RateLimiter()
{
    TimeLimiter timeconstraint = TimeLimiter.GetFromMaxCountByInterval(3, TimeSpan.FromSeconds(1));

    var tasks = new List<Task>();

    for (int i = 0; i < 15; i++)
    {
        tasks.Add(timeconstraint.Perform(ConsoleItAsync));
    }

    Task.WaitAll(tasks.ToArray());
}

private static async void ConsoleItAsync()
{
    await Task.Delay(800);

    var id = Thread.CurrentThread.ManagedThreadId;

    Trace.WriteLine($"{id}: {DateTime.Now:MM/dd/yyy HH:mm:ss.fff}");
}
```

**Comment by strohhut** [comment]
Hello @David-Desmaisons, it works the way you say when I cast to `Action`:

```
 tasks.Add(timeconstraint.Perform((Action)VoidConsoleItAsync));
```

I guess it run's concurrently because it's impossible to wait for the execution of `VoidConsoleItAsync` to finish so RateLimiter fires all the tasks allowed by maxCound and TimeSpan.

But then I guess the drawback also is that it impossible to really wait for all executions of `VoidConsoleItAsync` to finish because how can you wait when it doesn't return a task?

If I create an action via lambda it's sequential again because the lambda isn't async:
```
 tasks.Add(timeconstraint.Perform(() => TaskConsoleItAsync.Wait()));
```

So I guess you need an async void action that awaits somewhere at the beginning of the method to get concurrency with RateLimiter?

**Comment by David-Desmaisons** [comment]
@strohhut, I guess in your use case, you would want to wait for the dispatched task and not the task returned by rateLimiter.

Somethink like:

```C#
var tasks = new List<Task>();
Action dispatchTask = () => tasks.Add( ConsoleItAsync());
 for (int i = 0; i < 15; i++)
    {
        timeconstraint.Perform(dispatchTask);
    }
Task.WaitAll(tasks.ToArray());
```



```C#
private static async Task ConsoleItAsync()
{
    await Task.Delay(800);

    var id = Thread.CurrentThread.ManagedThreadId;

    Trace.WriteLine($"{id}: {DateTime.Now:MM/dd/yyy HH:mm:ss.fff}");
}
```

**Comment by strohhut** [comment]
Thank you for your help I'll try it.  But probably I'll use something simpler that offers concurrency out of the box like https://stackoverflow.com/a/44630243/764754

**Comment by hishhash2** [comment]
Hello @David-Desmaisons , If I want to pass some arguments into the dispatchTask method, how can I do this. 

```
public static async Task<IEnumerable<SubmitResp>> SendBatch(this SubmitClient client, IEnumerable<SubmitDetails> list, TimeLimiter timeLimiter)
{
	List<Task> tasks = new List<Task>();

	foreach (SubmitDetails sm in list)
	{
		tasks.Add(client.TrySubmitAsync(sm));
		//timeLimiter.Perform(dispatchTask(sm));
	}

	await Task.WhenAll(tasks).ConfigureAwait(false);

	return tasks.OfType<Task<SubmitResp>>().Select(x => x.Result);
}
```

Thanks
